cachito/web/content_manifest.py

```diff
@@ -232,3 +237,246 @@ def generate_icm(self, image_contents=None):
+        :param str version: the package's version
+        :param bool dev: if the package is only used in a dev environment
+        :param list[dict] dependencies: the package's dependencies as a list of JSON
+            dicts
+        :param str path: relative path to package from root of repository
+        :param str replaces: the dependency that is being replaced by this one in the
+            context of the request
+        """
+        self.name = name
+        self.type = type
+        self.version = version
+        self.dev = dev
+        self.path = path
+        self.replaces = replaces
+
+        self.dependencies = [
+            ContentManifestPackage.from_json(dependency) for dependency in dependencies
+        ]
+
+    def __repr__(self):
+        return (
+            f"<{self.__class__.__name__} name={self.name}, type={self.type}, "
+            f"version={self.version}>"
+        )
+
+    @property
+    def hash(self):
+        """Generate the package hash based on its name, type and version."""
+        return hash(f"{self.name}{self.type}{self.version}")
+
+    @classmethod
+    def validate_json(cls, package):
+        """
+        Validate the JSON representation of a package.
+
+        :param dict package: the JSON representation of a package
+        :raise ValidationError: if the JSON does not match the required schema
+        """
+        required = ("name", "type", "version")
+
+        if not isinstance(package, dict):
+            raise ValidationError("A package JSON must be an instance of dict")
+
+        for key in required:
+            if key not in package:
+                raise ValidationError(
+                    "A package must be a JSON object with the following "
+                    f'keys: {", ".join(sorted(required))}.'
+                )
+            if not isinstance(package[key], str):
+                raise ValidationError('The "{}" key of the package must be a string'.format(key))
+
+    @classmethod
+    def from_json(cls, package):
+        """
+        Create a Package object from JSON.
+
+        :param dict package: the dictionary representing the package
+        :return: the Package object
+        :rtype: ContentManifestPackage
+        """
+        cls.validate_json(package)
+        return cls(**package)
+
+    def to_json(self, dependencies=None, subpath=None):
+        """
+        Generate the JSON representation of the package.
+
+        :param list dependencies: the JSON representation of the dependencies associated with this
+            package
+        :param str subpath: relative path to package from root of repository
+        :return: the JSON form of the Package object
+        :rtype: dict
+        """
+        rv = {"name": self.name, "type": self.type, "version": self.version}
+        if dependencies is not None:
+            rv["dependencies"] = dependencies
+        if subpath:
+            rv["path"] = subpath
+        return rv
+
+    def to_purl(self):
```

Comment:
All the logic to generating the purls is copied and pasted from the Package model class, which will be removed in a following PR (CLOUDBLD-4731).

Reply:
The plan is slightly changed. :) The package and dependency model and data are not removed at this moment and will be kept for a period of time.

Reply:
@tkdchen could we remove the models? I'd like to keep only the db data but not the corresponding python models.

Reply:
@chmeliik Removing the models will not have corresponding database migrations to remove the table and data. Will this bring some potential issue?

Reply:
One issue would be before the tables are removed from the database, every time to generate database migration , alembic will detect the existing tables and generate new migration to remove them. That could make it inconvenient and error-prone probably.

Reply:
Hmm, that's true. I don't like the idea of keeping around dead code, but I also don't want to drop the data right away...

It seems that we might be able to get around that with some alembic magic: https://alembic.sqlalchemy.org/en/latest/cookbook.html#don-t-generate-any-drop-table-directives-with-autogenerate